### PR TITLE
Stop wrapping single changes in transactions where possible

### LINKED ISSUE #27439: Stop wrapping single changes in transactions where possible
We currently always wrap changes in a transaction - but that isn't necessary when only a single change is involved. Here's a comparison of a single change with and without a transaction:

|                                          Method | DatabaseType |       Mean |    Error |   StdDev |     Median | Ratio | RatioSD |
|------------------------------------------------ |------------- |-----------:|---------:|---------:|-----------:|------:|--------:|
|                           UpdateWithTransaction |     Postgres | 1,171.1 us | 20.53 us | 19.20 us | 1,172.1 us |  1.00 |    0.00 |
|                        UpdateWithoutTransaction |     Postgres |   998.9 us | 19.17 us | 16.99 us |   999.4 us |  0.85 |    0.02 |
| UpdateWithoutTransaction_SetImplicitTransaction |     Postgres |         NA |       NA |       NA |         NA |     ? |       ? |
|                                           Query |     Postgres |   202.0 us |  5.40 us | 15.93 us |   206.9 us |  0.18 |    0.01 |

So the difference on PostgreSQL is 14.6%. Note that since unneeded roundtrips are eliminated (2 on Npgsql, 3 on SQL Server and most other providers), the perf gain increases as server latency increases (the above 14.6% are against localhost, so it's a minimum figure).

On SQL Server (localhost) the gain is higher, 27.8%.

Note that when an external transaction already exists, we still need to create a savepoint, even if there's only one change, since databases behave quite differently when a failure occurs during transaction, and a rollback to a savepoint is necessary in at least some cases (e.g. PostgreSQL).

<details>
<summary>Benchmark code</summary>

```c#
BenchmarkRunner.Run<Benchmark>();

public class Benchmark
{
    private DbConnection _connection;
    private DbCommand _command;

    [Params(DatabaseType.Postgres, DatabaseType.SqlServer)]
    public DatabaseType DatabaseType { get; set; }

    private async Task Setup()
    {
        _connection = DatabaseType == DatabaseType.Postgres
            ? new NpgsqlConnection("Host=localhost;Username=test;Password=test;Max Auto Prepare=10")
            : new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Trust Server Certificate=true");
        await _connection.OpenAsync();

        using var cmd = _connection.CreateCommand();
        cmd.CommandText = @"
DROP TABLE IF EXISTS data;
CREATE TABLE data (id int PRIMARY KEY, num INT);
INSERT INTO data (id, num) VALUES (1, 1)";
        await cmd.ExecuteNonQueryAsync();
    }

    [GlobalSetup(Targets = new[] { nameof(UpdateWithTransaction), nameof(UpdateWithoutTransaction)})]
    public async Task Setup_Update()
    {
        await Setup();

        _command = _connection.CreateCommand();
        _command.CommandText = "UPDATE data SET num = num + 1";
    }

    [GlobalSetup(Target = nameof(UpdateWithoutTransaction_SetImplicitTransaction))]
    public async Task Setup_UpdateWithoutTransaction_SetImplicitTransaction()
    {
        if (DatabaseType != DatabaseType.SqlServer)
            throw new NotSupportedException("Benchmark relevant only for SQL Server");
        await Setup();

        _command = _connection.CreateCommand();
        _command.CommandText = "SET IMPLICIT_TRANSACTIONS OFF; UPDATE data SET num = num + 1";
    }


    [GlobalSetup(Target = nameof(Query))]
    public async Task Setup_Query()
    {
        await Setup();

        _command = _connection.CreateCommand();
        _command.CommandText = "SELECT num FROM data WHERE id = 1";
    }

    [Benchmark(Baseline = true)]
    public async Task UpdateWithTransaction()
    {
        var tx = _connection.BeginTransaction();
        _command.Transaction = tx;
        await _command.ExecuteNonQueryAsync();
        tx.Commit();
    }

    [Benchmark]
    public async Task UpdateWithoutTransaction()
    {
        await _command.ExecuteNonQueryAsync();
    }

    [Benchmark]
    public async Task UpdateWithoutTransaction_SetImplicitTransaction()
    {
        await _command.ExecuteNonQueryAsync();
    }

    [Benchmark]
    public async Task Query()
    {
        await _command.ExecuteNonQueryAsync();
    }
}

public enum DatabaseType
{
    Postgres,
    SqlServer
}
```

</details>

### PR DESCRIPTION
**Pull request by roji**
@AndriySvyryd here's a draft removing the unnecessary transaction where possible.

I got stuck in a thorny chicken and egg issue. To know whether a transaction needs to be started, I need to know if there's more than one batch. So the code as of now always peeks one batch forward; but this fails Can_insert_TPT_dependents_with_identity since GenerateColumnModifications gets called on ModificationCommand in the later batch, and the earlier batch hasn't yet been executed (and so hasn't propagated server-generated keys into the shared InternalEntityEntry).

Hopefully I'm understanding the situation right (am new here). One workaround could be to restart the enumeration if I see more than one batch; this would be optimized for the one-batch case, but would cause multiple enumerations (and so multiple TopologicalSort etc.) for multiple batches - not ideal.

Ideally, while preparing the second batch, we'd already know that the value will be propagated by the time we execute it, even if it hasn't yet been at the time we generate it; this way batch generation would no longer depend on execution order etc. But I have no idea how feasible/reasonable that is.

**NOTE:** Solved this, [see below](https://github.com/dotnet/efcore/pull/27500#issuecomment-1054386864).

Closes #27439

### CODE REVIEW

**Comment by roji** [comment]
@AndriySvyryd FYI I ended up solving the above by changing CommandBatchPreparer to allow checking if there is another batch up ahead, but without actually generating it; this makes CommandBatchPreparer itself behave like an enumerator, and unfortunately means we can't use yield.

Technically, this still isn't as good as being able to generate all the batches without executing any, since the current logic may return that there are batches, whereas when we actually get around to generate them, we find that there are no modifications to be done (i.e. update without write operations). In that case we'd get a transaction though we don't need, but that seems like an acceptable edge case.

Let me know what you think - this PR is close to ready, mostly test cleanup remains.

**Comment by AndriySvyryd** [comment]
I think we can avoid adding the peek API (and perhaps go back to using yield) if we signal that there are still more commands afterwards by adding a parameter to `ModificationCommandBatch.Complete` that will affect the value of `ModificationCommandBatch.RequiresTransaction`.

Note that `ModificationCommandBatch.Complete` will be removed by https://github.com/dotnet/efcore/issues/20664 since `ModificationCommandBatch.AddCommand` will always be called on the current batch before starting the next one.

**Comment by roji** [comment]
Login issues to SQL Server on Helix...

File: src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs [hunk 1]
```diff
@@ -45,86 +46,103 @@ public class SqlServerUpdateSqlGenerator : UpdateSqlGenerator, ISqlServerUpdateS
...
         var writeOperations = modificationCommands[0].ColumnModifications.Where(o => o.IsWrite).ToList();
         var keyOperations = modificationCommands[0].ColumnModifications.Where(o => o.IsKey).ToList();
 
         var defaultValuesOnly = writeOperations.Count == 0;
-        var nonIdentityOperations = modificationCommands[0].ColumnModifications
-            .Where(o => o.Property?.GetValueGenerationStrategy(table) != SqlServerValueGenerationStrategy.IdentityColumn)
+        var writableOperations = modificationCommands[0].ColumnModifications
+            .Where(o =>
+                o.Property?.GetValueGenerationStrategy(table) != SqlServerValueGenerationStrategy.IdentityColumn
+                && o.Property?.GetComputedColumnSql() is null
```
Comment (roji) [inline src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:69]: @AndriySvyryd this is the fix for #27507. I'm not sure I get the full logic below, so probably a good idea to give this special attention.

File: src/EFCore.Relational/Update/ReaderModificationCommandBatch.cs [hunk 2]
```diff
@@ -175,15 +212,33 @@ protected virtual void UpdateCachedCommandText(int commandPosition)
     protected virtual int GetParameterCount()
         => ModificationCommands.Sum(c => c.ColumnModifications.Count);
 
+    /// <inheritdoc />
+    public override void Complete()
+    {
+        UpdateCachedCommandText();
+
+        // Some database have a more where autocommit is off, and so executing a command outside of an explicit transaction implicitly
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Update/ReaderModificationCommandBatch.cs:220]:
```suggestion
        // Some database have a mode where autocommit is off, and so executing a command outside of an explicit transaction implicitly
```

File: src/EFCore.Relational/Update/UpdateSqlGenerator.cs [hunk 3]
```diff
@@ -104,9 +112,13 @@ protected virtual ISqlGenerationHelper SqlGenerationHelper
         {
             var keyOperations = operations.Where(o => o.IsKey).ToList();
 
+            requiresTransaction = true;
+
             return AppendSelectAffectedCommand(commandStringBuilder, name, schema, readOperations, keyOperations, commandPosition);
         }
 
+        requiresTransaction = false;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Update/UpdateSqlGenerator.cs:120]: Can move above

File: src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs [hunk 4]
```diff
@@ -45,86 +46,103 @@ public class SqlServerUpdateSqlGenerator : UpdateSqlGenerator, ISqlServerUpdateS
...
-            if (nonIdentityOperations.Count == 0
+            if (writableOperations.Count == 0
                 || readOperations.Count == 0)
             {
+                requiresTransaction = false;
                 foreach (var modification in modificationCommands)
                 {
-                    AppendInsertOperation(commandStringBuilder, modification, commandPosition);
+                    AppendInsertOperation(commandStringBuilder, modification, commandPosition, out var localRequiresTransaction);
+                    requiresTransaction = requiresTransaction || localRequiresTransaction;
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:82]: Shouldn't this also be `true` if there are more than 1 `modificationCommands` ?

File: test/EFCore.Relational.Specification.Tests/TestModels/SaveChangesScenariosModel/SaveChangesData.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.EntityFrameworkCore.TestModels.SaveChangesScenariosModel;
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Specification.Tests/TestModels/SaveChangesScenariosModel/SaveChangesData.cs:4]:
```suggestion
namespace Microsoft.EntityFrameworkCore.TestModels.StoreValueGeneration;
```

File: test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+        fixture.Reseed();
+
+        ClearLog();
+    }
+
+    #region Single operation
+
+    [ConditionalTheory]
+    [MemberData(nameof(IsAsyncData))]
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs:25]:
```suggestion
    [MemberData(
        nameof(DataGenerator.GetCombinations),
        new object[] { new[] { typeof(EntityState), typeof(GeneratedValues), typeof(bool) } },
        MemberType = typeof(DataGenerator))]
```

File: test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.EntityFrameworkCore.TestModels.SaveChangesScenariosModel;
+
+namespace Microsoft.EntityFrameworkCore.Update;
+
+#nullable enable
+
+public abstract class SaveChangesScenariosTestBase<TFixture> : IClassFixture<TFixture>
+    where TFixture : SaveChangesScenariosFixtureBase
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs:11]:
```suggestion
public abstract class StoreValueGenerationTestBase<TFixture> : IClassFixture<TFixture>
    where TFixture : StoreValueGenerationFixtureBase
```

File: src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:82]: Good batch, thanks...

Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:82]: I like the pun / freudian typo

File: test/EFCore.Relational.Specification.Tests/TestModels/SaveChangesScenariosModel/SaveChangesData.cs (hunk 5 above)
Comment (roji) [reply test/EFCore.Relational.Specification.Tests/TestModels/SaveChangesScenariosModel/SaveChangesData.cs:4]: Well, *actually* this also tests scenarios where there is no value generation :trollface: 

But am fine with whatever name.

Comment (AndriySvyryd) [reply test/EFCore.Relational.Specification.Tests/TestModels/SaveChangesScenariosModel/SaveChangesData.cs:4]: No value generation is also a value generation scenario :trollface: :trollface:

File: src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs (hunk 4 above)
Comment (roji) [reply src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:82]: I have been thinking too much about batching recently...

Comment (roji) [reply src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:82]: Fixed this. This currently isn't covered by the tests since they only do 2 inserts, which is below the 4 minimum limit. The next PR reduces the minimum to 1, so it will get covered there.

File: test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs (hunk 6 above)
Comment (roji) [reply test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs]: That's what I did originally - but it makes it harder to do SQL assertions for each scenario (and also somewhat harder to track exactly which scenario is being tested)... Let me know what you think.

Comment (roji) [reply test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs]: Also, it's harder to skip/override specific test cases if a provider wants to.

Comment (AndriySvyryd) [reply test/EFCore.Relational.Specification.Tests/Update/SaveChangesScenariosTestBase.cs]: It's fine

File: src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs [hunk 8]
```diff
@@ -45,86 +46,103 @@ public class SqlServerUpdateSqlGenerator : UpdateSqlGenerator, ISqlServerUpdateS
...
-                commandStringBuilder, modificationCommands, commandPosition, nonIdentityOperations, keyOperations, readOperations);
+                commandStringBuilder, modificationCommands, commandPosition, writableOperations, keyOperations, readOperations,
+                out requiresTransaction);
         }
 
         if (modificationCommands[0].Entries.SelectMany(e => e.EntityType.GetAllBaseTypesInclusive())
             .Any(e => e.IsMemoryOptimized()))
         {
-            if (!nonIdentityOperations.Any(o => o.IsRead && o.IsKey))
+            requiresTransaction = false;
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs:112]:
```suggestion
            requiresTransaction = modificationCommands.Count > 1;
```
